kliff/models/parameter.py

```diff
@@ -1,705 +1,430 @@
+    Modeled on `torch.nn.Parameters`, it inherits from `numpy.ndarray`. It is a numpy
+    array with additional attributes such as name, transform, etc. For maintaining compatibility,
+    use `get_numpy_array` (for getting a numpy array of parameters) and `get_opt_numpy_array`
+    (transformed numpy array but with only the optimizable values). `get_numpy_array` array
+    also helps in getting an instance of the Parameter to mutate safely, and pass it  to
+    any impure function.
+
+    Glossary:
+        - Model space: The space in which the model expects the parameters to be in. Currently,
+            all models in OpenKIM expect the parameters to be in the affine cartesian space.
+        - Parameter space: Parameter space is the space in which you want to sample/optimize
+            the parameters. Most often parameters are transformed using bijective transformations
+            of the model inputs for ease of optimization/sampling. For example, the log
+            transform is used in searching for the optimal parameters of sloppy model parameters.
+            There can be cases where transformation functions are not bijective, e.g. ceiling function for
+            mapping continuous parameters to discrete values. Parameter space is mostly
+            used for optimization, and not the model itself. If no transform_function is
+            provided then parameter space and model space are same.
+
+        All functions that needs input/output in the model space, will use the suffix
+        `_model_space` and `_param_space` for the transformed parameter space.
+
+    Attributes:
+        name : Name of the parameter.
+        transform : Instance of  ``ParameterTransform`` object to be applied to the parameter.
+        index : Index of the parameter in the parameter vector. used for setting the parameter
+            in the KIMPY.
+        bounds : Bounds for the parameter, must be numpy array of shape n x 2, with [n,0] as
+        lower bound, and [n,1] as the upper bound. If None, no bounds are applied.
+        opt_mask : A boolean or boolean array of the same shape as the parameter. For a
+            vector parameter ``opt_mask`` acts as a binary miask to determine which vector
+            components will be optimized, e.g. for a parameter with value [1., 2., 3.],
+            and opt_mask [True, False, True], only the first and third components will be
+            optimized, and second one will be presumed constant.
     """
 
-    def __init__(
-        self,
-        value: Union[Sequence[float], np.ndarray],
-        fixed: Optional[Sequence[bool]] = None,
-        lower_bound: Optional[Sequence[float]] = None,
-        upper_bound: Optional[Sequence[float]] = None,
-        name: Optional[str] = None,
-        index: Optional[int] = None,
+    name: str
+    transform_function: "ParameterTransform"
+    index: int
+    _is_transformed: bool
```

Comment:
Moved. they were there, as otherwise the static analyzer gives warning that "Attributes defined outside __init__". Does not change anything regarding the functionality.